src/devices/bus/epson_qx/option.h

```diff
@@ -0,0 +1,280 @@
+
+	auto out_inth1_callback() { return m_inth1_cb.bind(); }
+	auto out_inth2_callback() { return m_inth2_cb.bind(); }
+	template <unsigned Slot> auto out_intl_callback() { return m_intl_cb[Slot].bind(); }
+
+	auto out_drqf_callback() { return m_drqf_cb.bind(); }
+	template <unsigned Slot> auto out_drqs_callback() { return m_drqs_cb[Slot].bind(); }
+
+	auto out_rdyf_callback() { return m_rdyf_cb.bind(); }
+	auto out_rdys_callback() { return m_rdys_cb.bind(); }
+
+	memory_view::memory_view_entry& memview() const { return *m_view; }
+	address_space &iospace() const { return *m_iospace; }
+
+	void dackf_w(uint8_t data);
+	uint8_t dackf_r();
+	template <unsigned Slot> void dacks_w(uint8_t data) {  (*this)[Slot]->m_dmas_w_cb(data);  }
+	template <unsigned Slot> uint8_t dacks_r() { return (*this)[Slot]->m_dmas_r_cb(); }
+
+	template <void (option_slot_device::*slot_callback)(int)> void slots_w(int state) {
+		for (option_slot_device *slot : m_slot_list) {
+			(slot->*slot_callback)(state);
+		}
+	}
+
+	void set_inth1_line(uint8_t state, uint8_t slot);
+	void set_inth2_line(uint8_t state, uint8_t slot);
+	void set_intl_line(uint8_t state, uint8_t slot) { m_intl_cb[slot](state); }
+	void set_drqf_line(uint8_t state, uint8_t slot);
+	void set_drqs_line(uint8_t state, uint8_t slot) { m_drqs_cb[slot](state); }
+	void set_rdyf_line(uint8_t state, uint8_t slot);
+	void set_rdys_line(uint8_t state, uint8_t slot);
+
+	void add_slot(option_slot_device &slot);
+	option_slot_device* operator[](int index) const {assert(index < m_slot_list.size()); return m_slot_list[index]; }
+
+protected:
+	// device-level overrides
+	virtual void device_start() override;
+	virtual void device_reset() override;
+
+	required_address_space m_iospace;
+	memory_view::memory_view_entry *m_view;
+
+	devcb_write_line m_inth1_cb;
+	devcb_write_line m_inth2_cb;
+	devcb_write_line::array<5> m_intl_cb;
+
+	devcb_write_line m_drqf_cb;
+	devcb_write_line::array<4> m_drqs_cb;
+
+	devcb_write_line m_rdyf_cb;
+	devcb_write_line m_rdys_cb;
+
+	std::vector<option_slot_device *> m_slot_list;
+
+	uint8_t m_inth1;
+	uint8_t m_inth2;
+
+	uint8_t m_drqf;
+
+	uint8_t m_rdyf;
+	uint8_t m_rdys;
+};
+
+
+/* Epson Option Card interface */
+class device_option_expansion_interface : public device_interface
+{
+public:
+	void set_option_bus(option_bus_device &bus, int slot) { assert(!device().started()); m_bus = &bus; m_slot = slot; }
+
+protected:
+	device_option_expansion_interface(const machine_config &mconfig, device_t &device);
+
+	virtual void interface_pre_start() override;
+	option_slot_device* get_slot() { return (*m_bus)[m_slot]; }
+
+	option_bus_device  *m_bus;
+	int m_slot;
+};
+
+} // namespace bus::epson_qx
+
+
+// device type definition
+DECLARE_DEVICE_TYPE_NS(OPTION_BUS_SLOT, bus::epson_qx, option_slot_device)
+DECLARE_DEVICE_TYPE_NS(QX_OPTION_BUS, bus::epson_qx, option_bus_device)
+
+void option_bus_devices(device_slot_interface &device);
```

Comment:
The names `OPTION_BUS_SLOT`, `QX_OPTION_BUS` and `option_bus_devices` are still too generic to be throwing in the global namespace.  Change the device types to something like `EPSON_QX_OPTION_SLOT` and `EPSON_QX_OPTION_BUS` and put `option_bus_devices` inside `namespace bus::epson_qx`.